pandas/core/computation/align.py

```diff
@@ -49,7 +49,7 @@ def _align_core_single_unary_op(
 
 
 def _zip_axes_from_type(
-    typ: type[FrameOrSeries], new_axes: Sequence[Index]
+    typ: type[NDFrameT], new_axes: Sequence[Index]
```

Comment:
This is an unbound typevar? pyright not picking this up. can just use `Type[DataFrame | Series]` here?

Reply:
I think that pyright has some issues with nested `TypeVar(bound=...)`. I will create an issue.

Reply:
I replaced it with `Type[NDFrame]`. Similar to #42367, I can try to replace `NDFrame` with `DataFrame | Series` in private functions.

`List`, `Tuple`, ... are invariant (`Type` is not) which creates an interesting usecase for `TypeVar(bound=...)` to still allow sub-classes, see this example https://github.com/microsoft/pyright/issues/2344#issuecomment-927341908.

Reply:
> I replaced it with `Type[NDFrame]`. Similar to #42367, I can try to replace `NDFrame` with `DataFrame | Series` in private functions.

isn't it, replace `NDFrame` with `DataFrame | Series` in public functions? Although could also do this for private functions too and cast at the call site if needed.

> `List`, `Tuple`, ... are invariant (`Type` is not) which creates an interesting usecase for `TypeVar(bound=...)` to still allow sub-classes, see this example [microsoft/pyright#2344 (comment)](https://github.com/microsoft/pyright/issues/2344#issuecomment-927341908).

I'm not sure what the pyright error is, but mypy gives

```
/home/simon/t1.py:25: error: Argument 1 to "func1" has incompatible type "List[Bear]"; expected "List[Animal]"  [arg-type]
/home/simon/t1.py:25: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
/home/simon/t1.py:25: note: Consider using "Sequence" instead, which is covariant
```

The suggested fix is in line with https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#conventions ..

> avoid invariant collection types (list, dict) in argument positions, in favor of covariant types like Mapping or Sequence

I'm not sure how many functions we have that mutate the passed parameters where we can't do this.

We also have cases where where we have/will need overloads for tuple in indexing functions, so it may become necessary to use invariant collection types in argument positions in the overload.

Reply:
> isn't it, replace `NDFrame` with `DataFrame | Series` in public functions? Although could also do this for private functions too and cast at the call site if needed.

Yes, public functions that are not meant to take any NDFrame. I will do that later today.

I would be fine with using TypeVar in an invariant container (List) instead of using the covariant container (Sequence) with NDFrame. In both cases, it would probably be good to have a brief comment.

> I'm not sure what the pyright error is, but mypy gives
```
error: Argument of type "list[Bear]" cannot be assigned to parameter "a" of type "List[Animal]" in function "func1"
    TypeVar "_T@list" is invariant
      "Bear" is incompatible with "Animal" (reportGeneralTypeIssues)
```
reportGeneralTypeIssues is currently disabled for pandas.